File: native/android/policy/pro/thesquad/shell/StoreDestination.java
package pro.thesquad.shell;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Locale;

public final class StoreDestination {
    private final String origin;
    private final String hostname;

    private StoreDestination(String origin, String hostname) {
        this.origin = origin;
        this.hostname = hostname;
    }

    public static StoreDestination parse(String origin) {
        ParsedURL parsed = parseHTTPSURL(origin);
        if (parsed == null) {
            return null;
        }

        String path = parsed.uri.getRawPath();
        if (path != null && !path.isEmpty() && !path.equals("/")) {
            return null;
        }
        if (parsed.uri.getRawQuery() != null || parsed.uri.getRawFragment() != null) {
            return null;
        }

        String normalizedOrigin = "https://" + parsed.hostname;
        return new StoreDestination(normalizedOrigin, parsed.hostname);
    }

    public String origin() {
        return origin;
    }

    public String bootstrapURL() {
        return origin + "/api/app-distribution";
    }

    public boolean allows(String candidate) {
        ParsedURL parsed = parseHTTPSURL(candidate);
        return parsed != null && parsed.hostname.equals(hostname);
    }

    public boolean acceptsBootstrap(
            int status,
            String finalURL,
            boolean redirected,
            String distribution) {
        return status == 200 && !redirected && "store".equals(distribution)
            && bootstrapURL().equals(finalURL);
    }

    private static ParsedURL parseHTTPSURL(String value) {
        if (!passesRawInputChecks(value)) {
            return null;
        }

        final URI uri;
        try {
            uri = new URI(value);
        } catch (URISyntaxException exception) {
            return null;
        }

        String rawAuthority = uri.getRawAuthority();
        String rawHostname = uri.getHost();
        if (!uri.isAbsolute()
                || uri.isOpaque()
                || uri.getScheme() == null
                || !uri.getScheme().equalsIgnoreCase("https")
                || rawAuthority == null
                || rawAuthority.isEmpty()
                || !isASCII(rawAuthority)
                || rawAuthority.contains("%")
                || uri.getRawUserInfo() != null
                || rawHostname == null) {
            return null;
        }

        String hostname = rawHostname.toLowerCase(Locale.ROOT);
        int port = uri.getPort();
        if (!isValidDNSHostname(hostname) || (port != -1 && port != 443)) {
            return null;
        }

        String expectedAuthority = port == 443 ? hostname + ":443" : hostname;
        if (!rawAuthority.toLowerCase(Locale.ROOT).equals(expectedAuthority)) {
            return null;
        }

        return new ParsedURL(uri, hostname);
    }

    private static boolean passesRawInputChecks(String value) {
        if (value == null || value.isEmpty()) {
            return false;
        }
        int index = 0;
        while (index < value.length()) {
            int codePoint = value.codePointAt(index);
            if (Character.isWhitespace(codePoint)
                    || Character.isSpaceChar(codePoint)
                    || Character.getType(codePoint) == Character.CONTROL
                    || codePoint == '\\') {
                return false;
            }
            if (codePoint == '%') {
                if (index + 2 >= value.length()
                        || !isASCIIHexDigit(value.charAt(index + 1))
                        || !isASCIIHexDigit(value.charAt(index + 2))) {
                    return false;
                }
                index += 3;
                continue;
            }
            index += Character.charCount(codePoint);
        }
        return true;
    }

    private static boolean isASCII(String value) {
        for (int index = 0; index < value.length(); index++) {
            if (value.charAt(index) > 0x7F) {
                return false;
            }
        }
        return true;
    }

    private static boolean isValidDNSHostname(String hostname) {
        if (hostname.equals("localhost")
                || hostname.endsWith(".localhost")
                || isNumericIPAddress(hostname)) {
            return false;
        }

        String[] labels = hostname.split("\\.", -1);
        if (labels.length == 0) {
            return false;
        }
        for (String label : labels) {
            if (label.isEmpty()
                    || !isASCIIAlphanumeric(label.charAt(0))
                    || !isASCIIAlphanumeric(label.charAt(label.length() - 1))) {
                return false;
            }
            for (int index = 0; index < label.length(); index++) {
                char character = label.charAt(index);
                if (!isASCIIAlphanumeric(character) && character != '-') {
                    return false;
                }
            }
        }
        return true;
    }

    private static boolean isNumericIPAddress(String hostname) {
        String[] labels = hostname.split("\\.", -1);
        if (labels.length == 0) {
            return false;
        }
        for (String label : labels) {
            if (label.isEmpty()) {
                return false;
            }
            boolean decimal = true;
            for (int index = 0; index < label.length(); index++) {
                char character = label.charAt(index);
                if (character < '0' || character > '9') {
                    decimal = false;
                    break;
                }
            }
            boolean hexadecimal = label.length() > 2
                && label.charAt(0) == '0'
                && (label.charAt(1) == 'x' || label.charAt(1) == 'X')
                && isASCIIHexString(label, 2);
            if (!decimal && !hexadecimal) {
                return false;
            }
        }
        return true;
    }

    private static boolean isASCIIAlphanumeric(char character) {
        return (character >= '0' && character <= '9')
            || (character >= 'A' && character <= 'Z')
            || (character >= 'a' && character <= 'z');
    }

    private static boolean isASCIIHexDigit(int character) {
        return (character >= '0' && character <= '9')
            || (character >= 'A' && character <= 'F')
            || (character >= 'a' && character <= 'f');
    }

    private static boolean isASCIIHexString(String value, int startIndex) {
        for (int index = startIndex; index < value.length(); index++) {
            if (!isASCIIHexDigit(value.charAt(index))) {
                return false;
            }
        }
        return startIndex < value.length();
    }

    private static final class ParsedURL {
        private final URI uri;
        private final String hostname;

        private ParsedURL(URI uri, String hostname) {
            this.uri = uri;
            this.hostname = hostname;
        }
    }
}

File: native/ios/Policy/StoreDestination.swift
import Foundation

struct StoreDestination {
    let origin: String
    private let hostname: String

    init?(origin: String) {
        guard let parsed = Self.parseHTTPSURL(origin),
              parsed.components.percentEncodedPath.isEmpty
                || parsed.components.percentEncodedPath == "/",
              parsed.components.percentEncodedQuery == nil,
              parsed.components.percentEncodedFragment == nil else {
            return nil
        }

        self.origin = "https://\(parsed.hostname)"
        self.hostname = parsed.hostname
    }

    var bootstrapURL: String {
        "\(origin)/api/app-distribution"
    }

    func allows(_ candidate: String) -> Bool {
        guard let parsed = Self.parseHTTPSURL(candidate) else {
            return false
        }
        return parsed.hostname == hostname
    }

    func acceptsBootstrap(
        status: Int,
        finalURL: String,
        redirected: Bool,
        distribution: String?
    ) -> Bool {
        status == 200 && !redirected && distribution == "store"
            && finalURL == bootstrapURL
    }

    private struct ParsedURL {
        let components: URLComponents
        let hostname: String
    }

    private static func parseHTTPSURL(_ value: String) -> ParsedURL? {
        guard passesRawInputChecks(value),
              let rawAuthority = rawAuthority(in: value),
              rawAuthority.unicodeScalars.allSatisfy(\.isASCII),
              !rawAuthority.contains("%"),
              let components = URLComponents(string: value),
              components.url != nil,
              components.scheme?.lowercased() == "https",
              components.user == nil,
              components.password == nil,
              let rawHostname = components.host else {
            return nil
        }

        let hostname = rawHostname.lowercased()
        guard isValidDNSHostname(hostname),
              components.port == nil || components.port == 443 else {
            return nil
        }

        let expectedAuthority = components.port == 443
            ? "\(hostname):443"
            : hostname
        guard rawAuthority.lowercased() == expectedAuthority else {
            return nil
        }

        return ParsedURL(components: components, hostname: hostname)
    }

    private static func passesRawInputChecks(_ value: String) -> Bool {
        guard !value.isEmpty else {
            return false
        }

        let scalars = Array(value.unicodeScalars)
        var index = 0
        while index < scalars.count {
            let scalar = scalars[index]
            if scalar.value <= 0x20 || scalar.value == 0x7F || scalar.value == 0x5C {
                return false
            }
            if scalar.value == 0x25 {
                guard index + 2 < scalars.count,
                      isASCIIHexDigit(scalars[index + 1]),
                      isASCIIHexDigit(scalars[index + 2]) else {
                    return false
                }
                index += 2
            }
            index += 1
        }
        return true
    }

    private static func rawAuthority(in value: String) -> String? {
        guard let separator = value.range(of: "://") else {
            return nil
        }
        let authorityStart = separator.upperBound
        let authorityEnd = value[authorityStart...].firstIndex { character in
            character == "/" || character == "?" || character == "#"
        } ?? value.endIndex
        let authority = String(value[authorityStart..<authorityEnd])
        return authority.isEmpty ? nil : authority
    }

    private static func isValidDNSHostname(_ hostname: String) -> Bool {
        guard hostname != "localhost", !isNumericIPAddress(hostname) else {
            return false
        }

        let labels = hostname.split(separator: ".", omittingEmptySubsequences: false)
        guard !labels.isEmpty else {
            return false
        }

        return labels.allSatisfy { label in
            guard let first = label.unicodeScalars.first,
                  let last = label.unicodeScalars.last,
                  isASCIIAlphanumeric(first),
                  isASCIIAlphanumeric(last) else {
                return false
            }
            return label.unicodeScalars.allSatisfy { scalar in
                isASCIIAlphanumeric(scalar) || scalar.value == 0x2D
            }
        }
    }

    private static func isNumericIPAddress(_ hostname: String) -> Bool {
        let labels = hostname.split(separator: ".", omittingEmptySubsequences: false)
        return !labels.isEmpty && labels.allSatisfy { label in
            let scalars = Array(label.unicodeScalars)
            if scalars.allSatisfy({ $0.value >= 0x30 && $0.value <= 0x39 }) {
                return !scalars.isEmpty
            }
            return scalars.count > 2
                && scalars[0].value == 0x30
                && (scalars[1].value == 0x78 || scalars[1].value == 0x58)
                && scalars.dropFirst(2).allSatisfy(isASCIIHexDigit)
        }
    }

    private static func isASCIIAlphanumeric(_ scalar: Unicode.Scalar) -> Bool {
        (scalar.value >= 0x30 && scalar.value <= 0x39)
            || (scalar.value >= 0x41 && scalar.value <= 0x5A)
            || (scalar.value >= 0x61 && scalar.value <= 0x7A)
    }

    private static func isASCIIHexDigit(_ scalar: Unicode.Scalar) -> Bool {
        (scalar.value >= 0x30 && scalar.value <= 0x39)
            || (scalar.value >= 0x41 && scalar.value <= 0x46)
            || (scalar.value >= 0x61 && scalar.value <= 0x66)
    }
}
